### PR TITLE
Implements Tri-State checkboxes on Tailoring screen.

### LINKED ISSUE #112: Tailoring Screen: Checkboxes stay disabled after deselect all
Some checkboxes stay disabled after deselect all even if their parents and grandparents are active. To enable them again it's necessary to disable and re-enable their direct parents.

Steps:
step1
step2
step3
step4

Results:
result

### PR DESCRIPTION
**Pull request by WesleyCeraso**
Closes #112

Signed-off-by: Wesley Ceraso Prudencio <wcerasop@redhat.com>

### CODE REVIEW

**Comment by WesleyCeraso** [comment]
Also fixes #36, #23

**Comment by WesleyCeraso** [comment]
@openscap-jenkins retest this please

**Comment by rsprudencio** [comment]
@openscap-jenkins retest this please

**Comment by mpreisler** [comment]
This is pretty cool and seems to be working well. It is however very slow. Could we make it faster? I suspect it's probably the repeated wasted redraws that's making it slow.

If we can't make it faster we should at least show a busy cursor to let the user know we are doing something and the app hasn't crashed.

**Comment by mpreisler** [comment]
@yuumasato I would appreciate your review here as well. This is a big change.

**Comment by WesleyCeraso** [comment]
@mpreisler I  haven't noticed how slow it was when painting big trees. In fact it was really slow when a big part of the tree is expanded and you check/uncheck the most top item. It's now fixed.

File: src/TailoringWindow.cpp [hunk 1]
```diff
@@ -268,6 +268,7 @@ void TailoringWindow::synchronizeProfileItem()
 void TailoringWindow::synchronizeTreeItem()
 {
     synchronizeTreeItemSelections(mBenchmarkItem);
+    return;
```
Comment (mpreisler) [inline src/TailoringWindow.cpp]: ?

File: src/TailoringWindow.cpp [hunk 2]
```diff
@@ -315,6 +322,7 @@ void TailoringWindow::synchronizeTreeItemSelections(QTreeWidgetItem* treeItem)
                     treeItem->setCheckState(0, Qt::Unchecked);
                 }
                 break;
+
```
Comment (mpreisler) [inline src/TailoringWindow.cpp]: ?

**Comment by mpreisler** [comment]
It's faster now but when I select/deselect it screws up the scroll position and scrolls all the way up.

**Comment by mpreisler** [comment]
It's much better to just disable repainting while you change the tree. That's easier than collapsing and expanding parts of the tree.

**Comment by WesleyCeraso** [comment]
@mpreisler Yeah, I'll take take a better look at this.

**Comment by WesleyCeraso** [comment]
@mpreisler The problem was not exactly with the painting event, as it can be postponed by setUpdatesEnabled(bool enable), instead, if the treewidget was visible, for each data changed it would recalculate its geometry.
The only way I found to prevent it is turning the treewidget invisible. The last solution I gave (collapsing) did the same thing but would lost the scroll position as you noticed.

File: src/TailoringWindow.cpp [hunk 3]
```diff
@@ -300,160 +265,71 @@ void TailoringWindow::synchronizeProfileItem()
...
 
-void TailoringWindow::synchronizeTreeItem(QTreeWidgetItem* treeItem, struct xccdf_item* xccdfItem, bool recursive)
+void TailoringWindow::synchronizeTreeItem()
 {
-    ++mSynchronizeItemLock;
+    // setting visible as false seems to be the only way to avoid entering the "if" in
+    // https://github.com/qt/qt/blob/4.8/src/gui/itemviews/qabstractitemview.cpp#L3190
+    mUI.itemsTree->setVisible(false);
+    synchronizeTreeItemSelections(mBenchmarkItem);
+    mUI.itemsTree->setVisible(true);
```
Comment (yuumasato) [inline src/TailoringWindow.cpp]: Please, make it clearer that we are making it invisible and visible again due to perceived slowness when changing state of long trees.

**Comment by WesleyCeraso** [comment]
@yuumasato

**Comment by WesleyCeraso** [comment]
@yuumasato Empty groups are checked by default so that their parent can be checked, instead of partially checked, in case all its siblings are checked as well.

**Comment by yuumasato** [comment]
@WesleyCeraso That makes sense. 
These are great changes, thank you!
Ack.

**Pull request merged**